packages/mobile/src/fiatExchanges/BidaliScreen.tsx

```diff
@@ -0,0 +1,148 @@
+import { openUrl } from 'src/app/actions'
+import { bidaliPaymentRequested } from 'src/fiatExchanges/actions'
+import networkConfig from 'src/geth/networkConfig'
+import { celoTokenBalanceSelector } from 'src/goldToken/selectors'
+import i18n from 'src/i18n'
+import { emptyHeader } from 'src/navigator/Headers'
+import { navigate } from 'src/navigator/NavigationService'
+import { Screens } from 'src/navigator/Screens'
+import { TopBarTextButton } from 'src/navigator/TopBarButton'
+import { StackParamList } from 'src/navigator/types'
+import { stableTokenBalanceSelector } from 'src/stableToken/reducer'
+
+function useInitialJavaScript(
+  currency: CURRENCY_ENUM,
+  jsonBalances: string,
+  e164PhoneNumber: string | null
+) {
+  const [initialJavaScript, setInitialJavaScript] = useState<string | null>()
+  useEffect(() => {
+    if (initialJavaScript || !e164PhoneNumber) {
+      return
+    }
+
+    setInitialJavaScript(`
+      window.valora = {
+        paymentCurrency: "${CURRENCIES[currency].code.toUpperCase()}",
+        phoneNumber: "${e164PhoneNumber}",
+        balances: ${jsonBalances},
+        onPaymentRequest: function (paymentRequest) {
+          var payload = { method: 'onPaymentRequest', data: paymentRequest };
+          window.ReactNativeWebView.postMessage(JSON.stringify(payload));
+        },
+        openUrl: function (url) {
+          var payload = { method: 'openUrl', data: { url } };
+          window.ReactNativeWebView.postMessage(JSON.stringify(payload));
+        }
+      };
+      true; // note: this is required, or you'll sometimes get silent failures
+    `)
+  }, [jsonBalances, e164PhoneNumber, initialJavaScript])
+
+  return initialJavaScript
+}
+
+type RouteProps = StackScreenProps<StackParamList, Screens.BidaliScreen>
+type Props = RouteProps
+
+function BidaliScreen({ route }: Props) {
+  const onMessage = (event: WebViewMessageEvent) => {
+    const { method, data } = JSON.parse(event.nativeEvent.data)
+    switch (method) {
+      case 'onPaymentRequest':
+        const { amount, address, currency, description, chargeId } = data
+        const onPaymentSent = () => {
+          webViewRef.current?.injectJavaScript(`window.valora.paymentSent();`)
+        }
+        const onCancelled = () => {
+          webViewRef.current?.injectJavaScript(`window.valora.paymentCancelled();`)
+        }
+        dispatch(
+          bidaliPaymentRequested(
+            address,
+            amount,
+            currency,
+            description,
+            chargeId,
+            onPaymentSent,
+            onCancelled
+          )
+        )
+        break
+      case 'openUrl':
+        const { url } = data
+        dispatch(openUrl(url))
+        break
+    }
+  }
+
+  const webViewRef = useRef<WebView>(null)
+  const cusdBalance = useSelector(stableTokenBalanceSelector)
+  const celoBalance = useSelector(celoTokenBalanceSelector)
+  const jsonBalances = useMemo(
```

Comment:
Not expensive, but we want to avoid calling `injectJavaScript` when it didn't change.

Reply:
I think I see your point but wouldn't it be more direct to do that by changing the dependency array for the `injectJavascript` effect to include `cusdBalance` and `celoBalance` instead of using `jsonBalances`?

It's a bit of a nit but I don't think it's immediately obvious why it's being memoized (shallow object equality) and removing usage would just make it a little cleaner.

Reply:
`jsonBalances` is used in 2 different places, it's meant to be extensible with more tokens and changing a single place.